java/src/main/java/org/wildfly/openssl/OpenSSLClientSessionContext.java

```diff
@@ -70,34 +73,119 @@ public int getSessionCacheSize() {
-            // TODO (jrp) find a way to get the real host and port
             final ClientSessionKey key = new ClientSessionKey(host, port);
             // set with the session pointer from the found session
-            final ClientSessionInfo foundSessionPtr = clientSessions.get(key);
+            final ClientSessionInfo foundSessionPtr = getCacheValue(key);
             if (foundSessionPtr != null) {
                 SSL.getInstance().setSession(ssl, foundSessionPtr.session);
-                return foundSessionPtr.sessionId;
             }
         }
-        return null;
+    }
+
+    private void purgeOld() {
+        if (maxCacheSize > 0) {
+            final int removeSize = (cache.size() - maxCacheSize);
+            if (removeSize > 0) {
+                // Remove each entry until there are either no more entries or the size matches
+                for (int i = 0; i < removeSize; i++) {
+                    final CacheEntry oldest = accessQueue.poll();
+                    if (oldest != null) {
+                        removeCacheEntry(oldest.key());
+                    } else {
+                        // No need to continue as there are no more entries
+                        break;
+                    }
+                }
+            }
+        }
+    }
+
+    private void addCacheEntry(final ClientSessionKey key, final ClientSessionInfo newValue) {
+        CacheEntry value = cache.get(key);
+        if (value == null) {
+            value = new CacheEntry(key, newValue);
+            final CacheEntry result = cache.putIfAbsent(key, value);
+            if (result != null) {
+                value = result;
+                value.setValue(newValue);
+            }
+            bumpAccess(value);
+            if (maxCacheSize > 0 && cache.size() > maxCacheSize) {
+                // Remove the oldest entry
+                final CacheEntry oldest = accessQueue.poll();
+                if (oldest != value) {
+                    removeCacheEntry(oldest.key());
+                }
+            }
+        }
+    }
+
+    private ClientSessionInfo getCacheValue(final ClientSessionKey key) {
+        CacheEntry cacheEntry = cache.get(key);
+        if (cacheEntry == null) {
+            return null;
+        }
+        if (timeout > 0) {
+            long expires = cacheEntry.getTime() + (timeout * 1000);
+            if (System.currentTimeMillis() > expires) {
+                removeCacheEntry(key);
+                return null;
+            }
+        }
+
+        if (cacheEntry.hit() % 5 == 0) {
+            bumpAccess(cacheEntry);
+        }
+
+        return cacheEntry.getValue();
+    }
+
+    private ClientSessionInfo removeCacheEntry(final ClientSessionKey key) {
+        CacheEntry remove = cache.remove(key);
+        if (remove != null) {
+            Object old = remove.clearToken();
+            if (old != null) {
+                accessQueue.removeToken(old);
+            }
+            return remove.getValue();
```

Comment:
The session should be freed when it is removed from the session map. OpenSSL uses reference counting, when we get the session the reference count is increased, we need to make sure it is decreased when we are done with it